// Copyright 2024 The gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package chaotic_good_frame;

message Settings {
    // Connection id
    // - sent server->client on the control channel to specify the
    //   data channel connection id, one for each desired connection
    // - exactly one sent client->server on the data channel to 
    //   complete the connection
    repeated bytes connection_id = 1;
    // Flag true if this is a data channel (and not a control channel)
    bool data_channel = 2;
    // Requested alignment for the data channel
    // Client and server each send this with their preferences
    // Sent on data channel
    // Client sends with servers preferred alignment, and server sends
    // with the clients preferred alignment
    uint32 alignment = 3;
    // Max chunk size allowed to be sent to a peer (0 or omitted: no limit)
    // Sent on the control channel, applies to all messages.
    // If client sends zero, server must send zero or close the connection.
    uint32 max_chunk_size = 4;
    // Flow control support.
    // Sent on the control channel, if false server must send false or
    // close the connection.
    // If true, flow control is supported.
    bool flow_control = 5;
}

message UnknownMetadata {
    string key = 1;
    bytes value = 2;
}

message ClientMetadata {
    optional string path = 1;
    optional string authority = 2;
    optional uint64 timeout_ms = 3;

    repeated UnknownMetadata unknown_metadata = 100;
}

message ServerMetadata {
    optional uint32 status = 1;
    optional bytes message = 2;

    repeated UnknownMetadata unknown_metadata = 100;
}

message BeginMessage {
    uint64 length = 1;
}

message GoAway {
    uint32 last_stream_id = 1;
    uint32 status = 2;
    string message = 3;
}

message SolicitRequests {
    uint32 count = 1;
}

message StreamWindowUpdate {
    uint32 additional_bytes = 1;
}

message StreamInitialWindowUpdate {
    uint32 bytes = 1;
}

message TransportWindowUpdate {
    uint32 additional_bytes = 1;
}
